File: solution/1008. Construct Binary Search Tree from Preorder Traversal/1008_01.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    private TreeNode construct(int[] preorder, int start, int end) {
        
        int len = preorder.length;
        if (start == -1) {
            // System.out.println("build [null]");
            return null;
        }
        // System.out.println("build [start: " + start + ", end: " + end + "]");
        
        int val = preorder[start], left = -1, right = -1;
        for (int i = start+1; i <= end; i++) {
            if (preorder[i] < val) {
                left = i; break;
            }
        }
        for (int i  = start+1; i <= end; i++) {
            if (preorder[i] > val) {
                right = i; break;
            }
        }
        
        TreeNode root = new TreeNode(val);
        if (right != -1) {
            root.right = construct(preorder, right, end);
            root.left = construct(preorder, left, right-1);
        } else {
            root.left = construct(preorder, left, end);
        }
        return root;
    }
    
    public TreeNode bstFromPreorder(int[] preorder) {
        if (preorder.length == 0)
            return null;
        return construct(preorder, 0, preorder.length-1);
    }
}

File: solution/1008. Construct Binary Search Tree from Preorder Traversal/1008_01.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        n = preorder.size();
        return dfs(preorder, 0, n - 1);
    }
    
private:
    int n;
    TreeNode* dfs(vector<int>& preorder, int start, int end) {
        if (start > end) return nullptr;
        
        auto node = new TreeNode(preorder[start]);
        int i = start;
        while (i+1 <= end && preorder[i+1] < preorder[start]) ++i;
        node->left = dfs(preorder, start + 1, i);
        node->right = dfs(preorder, i + 1, end);
        return node;
    }
};


// 8 [5 1 7] [10 12]
